### ISSUE TITLE
Health check UI is not showing nor recording status history when using SQL server storage provider

### ISSUE THREAD

**Issue by AndrzejKroczynski** [Xabaril/AspNetCore.Diagnostics.HealthChecks#559]
**What happened**:
Health check UI is not showing nor recording status history when using SQL server storage provider.

**What you expected to happen**:
Health check UI shows and records history in the same way as when using in memory storage provider (which works just fine)

**How to reproduce it (as minimally and precisely as possible)**:
- create Web API project
- add health check UI as per code below
- start the project
- trigger status change on SampleCheck endpoint
- check if history is displayed, check if HealthCheckExecutionHistories is populated

**Source code sample**:
Startup.cs of .Net core API
```
 services
                .AddHealthChecksUI(setupSettings: setup =>
                {
                    setup.AddHealthCheckEndpoint("SampleCheck", "localhost:5111/status/ui");                    
                    setup.MaximumHistoryEntriesPerEndpoint(500);
                    setup.SetEvaluationTimeInSeconds(30);
                })
                .AddSqlServerStorage(Configuration.GetConnectionString("HealthChecksUIContext"));
```
Connection to db works fine, tables are created, other data is stored correctly, but HealthCheckExecutionHistories table remains empty.

**Anything else we need to know?**:

**Environment**:

- .NET Core version: 3.1
- Healthchecks version: 3.1.1
- Operative system: Widows 10
- Others:

**Comment by CarlosLanderas** [comment]
@AndrzejKroczynski the status history timeline only writes a new entry when the state switches from healthy to unhealthy and viceversa (so it is not writing continuously when the state is the current one). Are you aware of that?.

Are you trying this forcing the sql to be healthty-unhealthy-healthy?

As you can see here:

![image](https://user-images.githubusercontent.com/16475649/84327165-f00f8780-ab7e-11ea-9994-2f6c5f2633c3.png)

**Comment by AndrzejKroczynski** [comment]
@CarlosLanderas - yep - I am aware about that.

This works perfectly fine when I'm using in memory storage provider (this looks exactly as on screen you've provided).

The problem is than when I switch to SQL server storage provider it stops working so I only see the current status, but no time line showing how the status was changing (even though it actually changed coupe of times). There are also no records in HealthCheckExecutionHistories table (I guess this is where those changes should be kept).

So the summary is that if I use AddInMemoryStorage() all works fine, but with AddSqlServerStorage(...) history does not work (all other code is exactly the same).

**Comment by CarlosLanderas** [comment]
Thanks for the info! I'll take a look into it. It's really weird as the code running is the same for all providers

**Comment by CarlosLanderas** [comment]
@AndrzejKroczynski I am not able to reproduce. I am using the StorageProviders sample:

https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/tree/master/samples/HealthChecks.UI.StorageProviders

and everything is working fine:

![image](https://user-images.githubusercontent.com/16475649/84364324-393bf780-abd0-11ea-8f30-c922dca21967.png)

![image](https://user-images.githubusercontent.com/16475649/84365074-3ab9ef80-abd1-11ea-9bfb-c85370242234.png)


One question. Are you using the same sql instance for the SQL HealthCheck and the UI?. Maybe you are shutting down sql server for the healthcheck and the UI is not able to write as well?.

**Comment by AndrzejKroczynski** [comment]
@CarlosLanderas  - I'm using separate instance for Health Check UI.

Let me checkout the sample as well and see what is the difference between mine code and the sample.

**Comment by CarlosLanderas** [comment]
That sample runs local Healthchecks and UI. I am going to try to reproduce this using different processes

**Comment by AndrzejKroczynski** [comment]
@CarlosLanderas - checked the sample with local UI checks and it works perfectly fine, but when I switched to remote service (e.g. localhost:5111) it stopped working.... The JSON returned by both local and remote api seems to be in the same format...

**Comment by rajeshkandati** [comment]
Issue: Health check UI is not showing TimeLine and HealthCheckExecutionHistories table is empty.

Steps followed:
1) Created asp.net core 3.1 API and added AddHealthChecks under ConfigureServices as per below.
public static class HealthCheckExtensions
    {
        /// <summary>
        /// Add Health Checks
        /// </summary>
        /// <param name="services"></param>
        /// <param name="configuration"></param>
        /// <param name="env"></param>
        public static void AddHealthChecksService(this IServiceCollection services, IConfiguration configuration, IWebHostEnvironment env)
        {
            services.AddHealthChecks()
                .AddCheck<RandomHealthCheck>(env.ApplicationName.Trim());
        }

        /// <summary>
        /// Random Api check
        /// </summary>
        private class RandomHealthCheck : IHealthCheck
        {
            public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
            {
                //if (DateTime.UtcNow.Minute % 2 == 0)
                {
                    return Task.FromResult(HealthCheckResult.Healthy());
                }

                //return Task.FromResult(HealthCheckResult.Unhealthy(description: $"The healthcheck {context.Registration.Name} failed at minute {DateTime.UtcNow.Minute}"));
            }
        }
    }


2) Updated UseEndPoints as per below.

app.UseEndpoints(endpoints => {
                endpoints.MapControllers();
                endpoints.MapHealthChecks("/healthz", new HealthCheckOptions
                {
                    Predicate = r => r.Name.Equals(env.ApplicationName.Trim(), StringComparison.InvariantCultureIgnoreCase),
                    ResponseWriter = UIResponseWriter.WriteHealthCheckUIResponse
                });
            });

3) Created One more API and updated appsettings.json as per below
"HealthChecksUI": {
    "HealthChecks": [
      {
        "Name": "API1",
        "Uri": "https://API1.azurewebsites.net/healthz"
      }
    ],
    "Webhooks": [      
    ],
    "EvaluationTimeInSeconds": 10,
    "MinimumSecondsBetweenFailureNotifications": 60,
    //"HealthCheckDatabaseConnectionString": "Data Source=[PUT-MY-PATH-HERE]\\healthchecksdb",    
  }

4) Updated ConfigureServices(API 2) as per below
            services.AddHealthChecks();
              //adding health check UI services
            services.AddHealthChecksUI(setup =>
            {
                // Set the maximum history entries by endpoint that will be served by the UI api middleware
                //setup.MaximumHistoryEntriesPerEndpoint(50);
            }).AddSqlServerStorage(sConnection);

5) Updated UseEndPoints under Configure(API 2) as per below
 app.UseEndpoints(config =>
            {
                config.MapHealthChecks("/healthz", new HealthCheckOptions
                {
                    Predicate = _ => true,
                    ResponseWriter = UIResponseWriter.WriteHealthCheckUIResponse
                });

                config.MapHealthChecksUI(setup =>
                {
                    setup.UIPath = "/healthchecks-ui"; // this is ui path in your browser
                    setup.ApiPath = "/health-ui-api"; // the UI ( spa app )  use this path to get information from the store ( this is NOT the healthz path, is internal ui api )
                    setup.AddCustomStylesheet("wwwroot/healthcheck.css");
                });

                config.MapDefaultControllerRoute();
            });

Let me know if i am missing anything to persist executionhistories and to show Timeline.

**Comment by CarlosLanderas** [comment]
@AndrzejKroczynski @rajeshkandati 

I've created a sample from scratch using nuget packages and two different sql instances.
You can find the source code sample here:

https://github.com/CarlosLanderas/healthchecks-sample

Everything is working fine for me. Could you give it a try and compare with your codebase?.

Thanks!

![image](https://user-images.githubusercontent.com/16475649/84477597-27ba2480-ac90-11ea-8e6d-f8175e906f9a.png)

**Comment by AndrzejKroczynski** [comment]
@CarlosLanderas - Your sample with two sql instances works just fine, but I think it helped me to identify the problem.

I think the issue is that status history is only saved when OVERALL component status changes. For example if we have health check end point with 10 checks where one of the checks constantly fails, then overall status is constantly Unhealthy - but other 9 checks are still performed, but their history is not saved.

For example, if you replace sql check with:
```
  services
      .AddRouting()
      .AddHealthChecks()
      .AddCheck(name: "random",
              () =>
              {
                   return DateTime.UtcNow.Second % 2 == 0
                       ? HealthCheckResult.Healthy()
                        : HealthCheckResult.Unhealthy(description: "Test description here");
                })
       .AddCheck(name: "constant",
               () =>
               {
                   return 
                       HealthCheckResult.Unhealthy(description: "Test description here");
        });
```

In that case "random" check changes are not being tracked, because overall status remains unchanged.

**Comment by rajeshkandati** [comment]
@CarlosLanderas - Same here. Sample is working fine for me too. Is it possible to find if the HealthCheckEndpoint is available or not(Alive or dead) from the HealthCheckSample?

**Comment by AndrzejKroczynski** [comment]
@CarlosLanderas - were you able to reproduce this on your side?

**Comment by CarlosLanderas** [comment]
Hello @AndrzejKroczynski  and @rajeshkandati . Unfortunately I'm having very busy days. I'll try to check this tonight :). Thanks!

**Comment by woeterman94** [comment]
Can confirm that I have the same issue when using a postgress database as storage.

**Comment by xico002** [comment]
yup same problem with postgresql. Sometimes it saves history but other times it doesnt.

**Comment by xico002** [comment]
Anyone found any solution to this problem?

**Comment by CarlosLanderas** [comment]
@AndrzejKroczynski exactly, I need to check the code because this part was done by @unaizorrilla  but I think the intention of the status history is recording the Overall endpoint health status.

**Comment by AndrzejKroczynski** [comment]
@CarlosLanderas - thanks for answer.

The history icon is displayed next to each individual item, so it suggests that history tracking is per each individual item.

Anyways - from usability perspective it would be really nice if we track that on individual item level.

**Comment by CarlosLanderas** [comment]
Then we have two options. Moving de details icon to the endpoint header or saving history at the individual level.

@unaizorrilla what do you think?

**Comment by AndrzejKroczynski** [comment]
@CarlosLanderas @unaizorrilla - I would really appreciate if you decide to record that on individual level, because it adds usability in real-life scenario.

For example - I have web service with 10 health checks - one of health checks is constantly failing because third-party service has  issue. But the service that is failing is not critical, so I can live with that. In that case if another health check - this time critical is failing every now end then history will not show that because overall status doe not change.

**Comment by LeonT27** [comment]
Hello,

I'm having the same problem with history, when I have multiple checks configure and one is constantly changing to healthy and unhealthy, the history of that element is not present but, when I only leave one check in the configuration, the history is present. 

Like @AndrzejKroczynski said, I think is better for the user to record individual history of checks.

**Comment by netcorenewbee** [comment]
Hello, I have a similar issue when using SQLServer Storage. This table dbo.HealthCheckExecutionHistories contains history of only one endpoint out of 11 I have configured.

If the status changes for that one particular endpoint it is logged correctly, but the rest is omitted every single time.

**Comment by galaldev** [comment]
I have the same problem

**Comment by balayoglu** [comment]
Hi all,

Seems the problem hasn't been solved yet. I tried both InMemoryStorage and SqlServerStorage, but it doesn't show the historical changes on the statuses.

These are the package versions I have used:
    <PackageReference Include="AspNetCore.HealthChecks.SqlServer" Version="5.0.2" />
    <PackageReference Include="AspNetCore.HealthChecks.UI" Version="5.0.1" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.Client" Version="5.0.1" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.InMemory.Storage" Version="5.0.1" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.SqlServer.Storage" Version="5.0.1" />

**Comment by balayoglu** [comment]
Hi all,

Seems the problem hasn't been solved yet. I tried both InMemoryStorage and SqlServerStorage, but the UI doesn't display the historical changes on the statuses.

These are the package versions I have used:
```
    <PackageReference Include="AspNetCore.HealthChecks.SqlServer" Version="5.0.2" />
    <PackageReference Include="AspNetCore.HealthChecks.UI" Version="5.0.1" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.Client" Version="5.0.1" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.InMemory.Storage" Version="5.0.1" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.SqlServer.Storage" Version="5.0.1" />
```

**Comment by rahul230691** [comment]
Does recording status history timeline not work with following packages? It shows me with "Healthy" or "Unhealthy". No history graph.

<ItemGroup>
    <PackageReference Include="AspNetCore.HealthChecks.AzureServiceBus" Version="5.1.1" />
    <PackageReference Include="AspNetCore.HealthChecks.Redis" Version="5.0.2" />
    <PackageReference Include="AspNetCore.HealthChecks.UI" Version="3.1.3" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.Client" Version="3.1.2" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.InMemory.Storage" Version="3.1.2" />
    <PackageReference Include="AspNetCore.HealthChecks.UI.SqlServer.Storage" Version="3.1.2" />
    <PackageReference Include="AspNetCore.HealthChecks.Uris" Version="5.0.1" />
  </ItemGroup>

**Comment by s4lvo** [comment]
Hi all,
has anyone found a solution?
Seems the problem hasn't been solved yet

**Comment by OneCrazyRussian** [comment]
I am having the same issue with latest version and on NET 6

First my thought was that it's the custom check service that does not do something so I've fallen back to premade Sql server checker

```

//builder.Services.AddHealthChecks().AddCheck<DbHealthCheckService>("Database connection");
builder.Services.AddHealthChecks().AddSqlServer(builder.Configuration.GetConnectionString("[REDACTED]"),name: "[REDACTED] DB");
builder.Services.AddHealthChecks().AddSqlServer(builder.Configuration.GetConnectionString("[REDACTED]"), name: "[REDACTED] DB");

//setup health check UI
builder.Services.AddHealthChecksUI(options =>
{
    options.SetEvaluationTimeInSeconds(60); //Sets the time interval in which HealthCheck will be triggered
    options.MaximumHistoryEntriesPerEndpoint(20); //Sets the maximum number of records displayed in history
    options.AddHealthCheckEndpoint("Health Checks API", "/health"); //Sets the Health Check endpoint
}).AddInMemoryStorage(); //Here is the memory bank configuration

...

app.MapControllers();

app.MapHealthChecks("/health", new HealthCheckOptions
{
    Predicate = _ => true,
    ResponseWriter = UIResponseWriter.WriteHealthCheckUIResponse
});

app.UseHealthChecksUI(options => { options.UIPath = "/dashboard"; });
```

But both premade SQL check and custom one are returning empty history array when you request /healthchecks-api

```
[
	{
		"id": 1,
		"status": "Healthy",
		"onStateFrom": "2022-06-24T15:02:16.8946627+03:00",
		"lastExecuted": "2022-06-24T15:09:17.8045469+03:00",
		"uri": "/health",
		"name": "Health Checks API",
		"discoveryService": null,
		"entries": [
			{
				"id": 1,
				"name": "[REDACTED] DB",
				"status": "Healthy",
				"description": null,
				"duration": "00:00:00.0252148",
				"tags": []
			},
			{
				"id": 2,
				"name": "[REDACTED] DB",
				"status": "Healthy",
				"description": null,
				"duration": "00:00:00.0250488",
				"tags": []
			}
		],
		"history": []
	}
]
```

**Comment by marcasmar94** [comment]
Any updates on this ?

**Comment by sungam3r** [comment]
Nope. See #1714

**Comment by jsanjuan2016** [comment]
As workaround I ended up creating a trigger on the database (in my case Sqlite):

CREATE TRIGGER after_update_executions 
   AFTER UPDATE ON Executions
   WHEN old.Status <> new.Status
BEGIN
	INSERT INTO HealthCheckExecutionHistories ([Name], [Status], [On], [HealthCheckExecutionId])
	VALUES
	(
		new.Name,
		new.Status,
		new.OnStateFrom,
		new.id
	);
END

**Comment by JonasJes** [comment]
@jsanjuan2016 I converted it to SQL and tried it out.
While I got entries in the `HealthCheckExecutionHistories`, nothing appeared in the UI.
I think you are triggering and logging from the wrong table.
Instead of `Executions` you should use `HealthCheckExecutionEntries` as the name in `HealthCheckExecutionHistories` should match the health check entries

This worked for me as a temporary workaround:
```sql
CREATE TRIGGER after_update_executionentries ON HealthCheckExecutionEntries
   AFTER UPDATE
AS DECLARE @Name NVARCHAR(500), @Status INT, @On DATETIME2(7), @HealthCheckExecutionId INT, @Description NVARCHAR(MAX)
SELECT @Name = ins.Name FROM INSERTED ins;
SELECT @Status = ins.Status FROM INSERTED ins;
SELECT @On = GETDATE();
SELECT @HealthCheckExecutionId = ins.HealthCheckExecutionId FROM INSERTED ins;
SELECT @Description = ins.[Description] FROM INSERTED ins;
BEGIN
    SET NOCOUNT ON;
    IF UPDATE([Status])
    BEGIN
        INSERT INTO HealthCheckExecutionHistories
            ([Name], [Status], [On], [HealthCheckExecutionId], [Description])
        VALUES
            (
                @Name,
                @Status,
                @On,
                @HealthCheckExecutionId,
                @Description
	    );
        PRINT 'HealthCheckExecutionEntries was updated and an event trigger was updated into HealthCheckExecutionHistories'
    END
END
```

**Comment by DeepWorksStudios** [comment]
i can comfirm this is still an issue today with the lastest versions

**Comment by Luk164** [comment]
Same issue for me running a docker container version with PosgreSQL 15 and running within my .net9 app with in memory storage

**Comment by s-beji** [comment]
Same issue with .net 10 preview
